plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/HiveMetastoreTableOperations.java

```diff
@@ -97,7 +99,15 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
             }
         }
         finally {
-            thriftMetastore.releaseTableLock(lockId);
+            try {
+                thriftMetastore.releaseTableLock(lockId);
```

Comment:
Consider throwing a `CommitStateUnknownException` in case of dealing with an exception.

Reply:
@findinpath 

I tried that locally in the test case, it has inserted the data successfully but throws  `CommitStateUnknownException`. To prove that I caught the `Exception` and selected the data to validate.

this
https://github.com/osscm/trino/blob/releaselock-handle-error/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/hms/TestIcebergHiveMetastoreTableOperationsReleaseLockFailure.java#L109-L110

modified to:

```
public void testReleaseLockFailureDoesNotCorruptTheTable()
    {
        String tableName = "test_release_lock_failure";
        query(format("CREATE TABLE %s (a_varchar) AS VALUES ('Trino')", tableName));
        try {
            query(format("INSERT INTO %s VALUES 'rocks'", tableName));
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        assertQuery("SELECT * FROM " + tableName, "VALUES 'Trino', 'rocks'");
        try {
            query(format("INSERT INTO %s VALUES 'all the time'", tableName));
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        assertQuery("SELECT * FROM " + tableName, "VALUES 'Trino', 'rocks', 'all the time'");
    }
```

Reply:
@findinpath one thing to consider is that if the table update succeeded but we throw an exception due to the unlocking error, that would end up overall failing the query. If Trino runs in a pipeline which automatically triggers a query retry upon a failure, that could result in duplicated data for example

Reply:
Would wrapping commitstateunknownexception in a trinoexception with a specific trino-iceberg error code on which the user can react to make sense ?

Reply:
See https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueTableOperationsInsertFailure.java#L135-L137 

as an alternative to do explicitly try-catch within the test code.

Reply:
cc @RussellSpitzer @aokolnychyi
wondering if you have any thoughts on handling the `releaseLock` issue as well. thanks!

Reply:
> with a specific trino-iceberg error code on which the user can react to

I think this makes sense if we do decide to throw an exception and fail the query. But I think the overall question is to decide which tradeoff is more acceptable for us:
- Successful updates don't fail due to unlock problems, and therefore have no side effects in a scripted/orchestrated environment (like data duplication due to rerun). However, in rare cases when the lock release fails it only logs a warning, meaning that users are not aware that for a (probably short) time period their commits might hang
- Successful updates sometimes fail (because of a lock release problem), which can have unintended consequences depending on the execution environment. However, the user becomes aware of the lock release problem and potential subsequent update problems that could result from that and can take manual action

This is an Iceberg-wide problem, so besides this tradeoff, it might be interesting to consider what other engines are doing in the same scenario so that we consider interoperability. By using HiveTableOperations, Spark, Flink, Hive, Impala, etc. swallow the exception and opt for choice #1.

Reply:
Given that we handle anyway with `CommitStateUnknownException` the `replaceTable` I think we can as well use it for `releaseLock` operation failures.

Reply:
I think `releaseLock` is not related to the commit state. Consider the case we know for sure that the `replaceTable` succeeded, and only the subsequent `releaseLock` failed. In that case the commit state is actually known (i.e. successful), so throwing this exception would not be appropriate IMO

Reply:
Now we have 2 options 
1. Catch `Exception` and log and do not propagate. PR's current status.
2. Catch `Exception` and throw `CommitStateUnknownException` to the `BaseTransaction` 

I feel option 1 seems to be aligned with the Spark and Flink, though as the `releaseLock` failure won't be quite often so it won't be that bad. 
The only problem I see is that when users see this error, they have to check the table to see if it's in the correct state and then Ok. So, might not be deterministic for the user.

IMP, If the locking also would have been be able to handled by the Iceberg's API, then it would have been ok, but as its being handled by the engines, it is difficult to manage it using `CommitStateUnknownException`